contracts/peg/NonCustodialPSM.sol

```diff
@@ -0,0 +1,460 @@
+            "PegStabilityModule: Mint fee exceeds max fee"
+        );
+        uint256 _oldMintFee = mintFeeBasisPoints;
+        mintFeeBasisPoints = newMintFeeBasisPoints;
+
+        emit MintFeeUpdate(_oldMintFee, newMintFeeBasisPoints);
+    }
+
+    /// @notice internal helper function to set the redemption fee
+    function _setRedeemFee(uint256 newRedeemFeeBasisPoints) internal {
+        require(
+            newRedeemFeeBasisPoints <= MAX_FEE,
+            "PegStabilityModule: Redeem fee exceeds max fee"
+        );
+        uint256 _oldRedeemFee = redeemFeeBasisPoints;
+        redeemFeeBasisPoints = newRedeemFeeBasisPoints;
+
+        emit RedeemFeeUpdate(_oldRedeemFee, newRedeemFeeBasisPoints);
+    }
+
+    /// @notice helper function to set the surplus target
+    function _setPCVDeposit(IPCVDeposit newSurplusTarget) internal {
+        require(
+            address(newSurplusTarget) != address(0),
+            "PegStabilityModule: Invalid new surplus target"
+        );
+        IPCVDeposit oldTarget = pcvDeposit;
+        pcvDeposit = newSurplusTarget;
+
+        emit PCVDepositUpdate(oldTarget, newSurplusTarget);
+    }
+
+    // ----------- Public State Changing API -----------
+
+    /// @notice function to redeem FEI for an underlying asset
+    /// We do not burn Fei; this allows the contract's balance of Fei to be used before the buffer is used
+    /// In practice, this helps prevent artificial cycling of mint-burn cycles and prevents a griefing vector.
+    /// This function will deplete the buffer for pulling from the PCV Deposit
+    function redeem(
+        address to,
+        uint256 amountFeiIn,
+        uint256 minAmountOut
+    )
+        external
+        virtual
+        override
+        nonReentrant
+        whenNotPaused
+        whileRedemptionsNotPaused
+        returns (uint256 amountOut)
+    {
+        updateOracle();
+
+        amountOut = _getRedeemAmountOut(amountFeiIn);
+        require(
+            amountOut >= minAmountOut,
+            "PegStabilityModule: Redeem not enough out"
+        );
+        _depleteBuffer(amountOut);
+
+        IERC20(fei()).safeTransferFrom(msg.sender, address(this), amountFeiIn);
+
+        _transfer(to, amountOut);
+
+        emit Redeem(to, amountFeiIn, amountOut);
+    }
+
+    /// @notice function to buy FEI for an underlying asset
+    /// We first transfer any contract-owned fei, then mint the remaining if necessary
+    function mint(
+        address to,
+        uint256 amountIn,
+        uint256 minAmountOut
+    )
+        external
+        virtual
+        override
+        nonReentrant
+        whenNotPaused
+        whileMintingNotPaused
+        returns (uint256 amountFeiOut)
+    {
+        updateOracle();
+
+        amountFeiOut = _getMintAmountOut(amountIn);
+        require(
+            amountFeiOut >= minAmountOut,
+            "PegStabilityModule: Mint not enough out"
+        );
+
+        _transferFrom(msg.sender, address(this), amountIn);
```

Comment:
transfer to the PCVDeposit? Maybe don't call deposit to save gas

Reply:
that would be a pretty massive gas saver if we didn't call deposit

Reply:
done

Reply:
Is this done? Still looks like it's transferring to the psm?

Reply:
Also, if we do this, we need to have some kind of threshold for calling deposit() on the pcv deposit automatically, else it would be relatively cheap for an actor to mint/redeem cycle and pull all of our assets out of the underlying yield-earning strategy in the pcv deposit

Reply:
@ElliotFriedman why did you resolve this comment?